src/middleware/streamingReceiver.js

```diff
@@ -0,0 +1,364 @@
+import { Types } from 'mongoose'
+import * as auditing from '../auditing'
+import { genAuthAudit } from './authorisation'
+import * as matching from './requestMatching'
+
+config.authentication = config.get('authentication')
+
+let bucket
+
+function streamingReceiver (ctx, statusEvents) {
+  let counter = 0
+  let size = 0
+
+  if (!bucket) {
+    bucket = getGridFSBucket()
+  }
+
+  ctx.state.downstream = new Readable()
+  ctx.state.downstream._read = () => {}
+
+  let gridFsStream
+
+  if (statusEvents && statusEvents.startRequest) {
+    statusEvents.startRequest(ctx.request.headers)
+  }
+
+  /*
+  * Only transactions that were requested to be rerun should have this
+  * custom header (the GridFS fileId of the body for this transaction)
+  */
+  const bodyId = ctx.request.headers['x-body-id']
+
+  const storeRequestBody = (['POST', 'PUT', 'PATCH'].includes(ctx.req.method)) && ctx.authorisedChannel.requestBody
+
+  if (storeRequestBody) {
+    if (!bodyId) {
+      /*
+      *   Request has a body, so stream it into GridFs
+      */
+      gridFsStream = bucket.openUploadStream()
+
+      ctx.requestTimestamp = new Date()
+
+      // Get the GridFS file object that was created
+      ctx.request.bodyId = gridFsStream.id
+
+      // Create the transaction for Request (started receiving)
+      // Side effect: Updates the Koa ctx with the transactionId
+      ctx.state.requestPromise = messageStore.initiateRequest(ctx)
+
+      gridFsStream
+        .on('error', (err) => {
+          if (statusEvents && statusEvents.gridFsError) {
+            statusEvents.gridFsError(err, ctx.request.bodyId)
+          }
+        })
+    } else {
+      /*
+      *   Request is a rerun, therefore has a bodyId, but no body.
+      *      So, stream the body from GridFs and send it downstream
+      */
+      const fileId = new Types.ObjectId(bodyId)
+      gridFsStream = bucket.openDownloadStream(fileId)
+
+      ctx.request.bodyId = fileId
+      ctx.state.requestPromise = messageStore.initiateRequest(ctx)
+
+      gridFsStream
+        .on('data', (chunk) => {
+          ctx.req.push(chunk)
+        })
+        .on('end', () => {
+          logger.info(`** END OF INPUT GRIDFS STREAM **`)
+          ctx.req.push(null)
+        })
+        .on('error', (err) => {
+          if (statusEvents && statusEvents.gridFsError) {
+            statusEvents.gridFsError(err, bodyId)
+          }
+        })
+    }
+  } else {
+    /*
+    *  GET and DELETE come in here to persist the initial request transaction
+    */
+    ctx.state.requestPromise = messageStore.initiateRequest(ctx)
+  }
+
+  ctx.req
+    .on('data', (chunk) => {
+      counter++;
+      size += chunk.toString().length
+      logger.info(`Read request CHUNK # ${counter} [ Total size ${size}]`)
+
+      // Write chunk to GridFS & downstream
+      if (storeRequestBody && !bodyId) {
+        gridFsStream.write(chunk)
+      }
+
+      ctx.state.downstream.push(chunk)
+    })
```

Comment:
Reading all the data in chunk out of a stream and into another stream doesn't seem necessary, we could just us `.pipe()` to pipe this into the gridFsStream and then assign the `ctx.req` stream to `ctx.state.downstream` rather than copy chunk over one at a time.